x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/netty4/SecurityNetty4Transport.java

```diff
@@ -151,15 +160,44 @@ protected ChannelHandler getClientChannelInitializer(DiscoveryNode node, Connect
 
     @Override
     protected InboundPipeline getInboundPipeline(boolean isRemoteClusterServerChannel) {
-        return new InboundPipeline(
-            getStatsTracker(),
-            threadPool::relativeTimeInMillis,
-            isRemoteClusterServerChannel
-                ? new InboundDecoder(recycler, RemoteClusterPortSettings.MAX_REQUEST_HEADER_SIZE.get(settings))
-                : new InboundDecoder(recycler),
-            new InboundAggregator(getInflightBreaker(), getRequestHandlers()::getHandler, ignoreDeserializationErrors()),
-            this::inboundMessage
-        );
+        if (false == isRemoteClusterServerChannel) {
+            return super.getInboundPipeline(false);
+        } else {
+            return new InboundPipeline(
+                getStatsTracker(),
+                threadPool::relativeTimeInMillis,
+                new InboundDecoder(recycler, RemoteClusterPortSettings.MAX_REQUEST_HEADER_SIZE.get(settings), false),
+                new InboundAggregator(getInflightBreaker(), getRequestHandlers()::getHandler, ignoreDeserializationErrors()),
+                this::inboundMessage
+            ) {
+                @Override
+                protected void headerReceived(TcpChannel channel, Header header) {
+                    if (header.isHandshake() == false) {
+                        // eagerly (before buffering the full request) authenticate all request headers for this type of channel
+                        assert header.isRequest();
+                        // authn is mostly async, avoid buffering anymore data while authn is in progress
+                        boolean isNettyChannel = channel instanceof Netty4TcpChannel;
+                        if (isNettyChannel) {
+                            ((Netty4TcpChannel) channel).getNettyChannel().config().setAutoRead(false);
+                        }
+                        // this prevents thread-context changes to propagate beyond the validation, as netty worker threads are reused
+                        try (ThreadContext.StoredContext ignore = threadPool.getThreadContext().newStoredContext()) {
+                            crossClusterAccessAuthenticationService.tryAuthenticate(
+                                header.getRequestHeaders(),
+                                ActionListener.wrap(aVoid -> {
+                                    // authn is successful -> NOOP (the complete request will be subsequently authn & authz & audited)
+                                    if (isNettyChannel) {
+                                        ((Netty4TcpChannel) channel).getNettyChannel().config().setAutoRead(true);
+                                    }
+                                }, e -> uncaughtExceptionReference.set(new HeaderValidationException(header, e)))
```

Comment:
Option 2 allows the client to control the flow here by never sending the final components of the message. So I'm not sure that is a great option.

fire exception caught probably works (it will dispatch to the event loop).

Also - I want to flag that this current design potentially results in multiple responses for a single request. It is possible that we have responded on the normal auth pathway along with responding on this specific pathway. That means the second response will throw some type of error on the receiving node when it attempts to look for an already removed response hander. I'm not sure if that matters.

Reply:
If we go with fire exception caught, I think we don't need the InboundPipeline change?

Reply:
> I want to flag that this current design potentially results in multiple responses for a single request

Thanks for flagging it. I think this is fine for now considering (1) the connection is ultimately closed; (2) the behaviour is limited to the new RCS server channel. Also, unhandled responses are just ignored and do not trigger exceptions on the client side

https://github.com/elastic/elasticsearch/blob/f307e6fedde203824431c12045db89089d923b32/server/src/main/java/org/elasticsearch/transport/InboundHandler.java#L120-L123

We definitely should work on removing the double authentication pattern which will also fix this behaviour. But I think it can be a future enhancement outside of this release (8.10.0).

Reply:
> If we go with fire exception caught, I think we don't need the InboundPipeline change?

Yeah. That's what we were considering and what Yang brought up in https://github.com/elastic/elasticsearch/pull/98814#discussion_r1303676935 .

> fire exception caught probably works (it will dispatch to the event loop). although it does feel like an unusual api usage.

I agree. Calling that API on a pipeline that's (potentially) not currently processing anything at that time seems unusual to me too, but I think it works (though it's the callers responsability to call it from the eventloop thread, I think).

The `InboundPipeline`s support for not accepting anymore bytes if an exception was thrown during some previous processing seems like a better fit, design wise, to me.

But since I see the sentiment here is for the other alternative, I'll go with that.

Reply:
Pushed [ee41575](https://github.com/elastic/elasticsearch/pull/98814/commits/ee41575a4d905a7eece2c92eb14ff5885a64e681)